llvm/lib/SandboxIR/SandboxIR.cpp

```diff
@@ -262,6 +262,113 @@ const char *Instruction::getOpcodeName(Opcode Opc) {
   llvm_unreachable("Unknown Opcode");
 }
 
+llvm::Instruction *Instruction::getTopmostLLVMInstruction() const {
+  Instruction *Prev = getPrevNode();
+  if (Prev == nullptr) {
+    // If at top of the BB, return the first BB instruction.
+    return &*cast<llvm::BasicBlock>(getParent()->Val)->begin();
+  }
+  // Else get the Previous sandbox IR instruction's bottom IR instruction and
+  // return its successor.
+  llvm::Instruction *PrevBotI = cast<llvm::Instruction>(Prev->Val);
+  return PrevBotI->getNextNode();
+}
+
+BBIterator Instruction::getIterator() const {
+  auto *I = cast<llvm::Instruction>(Val);
+  return BasicBlock::iterator(I->getParent(), I->getIterator(), &Ctx);
+}
+
+Instruction *Instruction::getNextNode() const {
+  assert(getParent() != nullptr && "Detached!");
+  assert(getIterator() != getParent()->end() && "Already at end!");
+  auto *LLVMI = cast<llvm::Instruction>(Val);
+  assert(LLVMI->getParent() != nullptr && "LLVM IR instr is detached!");
+  auto *NextLLVMI = LLVMI->getNextNode();
```

Comment:
what about multi-instruction instructions?

Reply:
So here is how this works: `NextLLVMI` is an LLVM instruction that belongs to the next SandboxIR instruction, whether it's a single-Instruction or multi-Instruction. If it's a multi-Instruction, then `NextLLVMI` will be the topmost LLVM instruction. In either case `Ctxt.getValue(NextLLVMI)` returns the correct SandboxIR Instruction that maps to `NextLLVMI`. Does this make sense?

Reply:
I meant if `this` is a multi-instruction, then `NextLLVMI` will point to the second LLVM instruction in `this`. Or are you going to override `getNextNode()` in the multi-instruction subclass?

Reply:
Ah the confusing part is that `Val` always points to the bottom LLVM Instruction of a `sandboxir::Instruction`, not the topmost one.
So if `this` is a multi-instruction, `LLVMI` is the bottom-most LLVM Instruction in `this`, so `LLVMI->getNextNode()` points to an LLVM instruction that does **not** belong to `this`, but instead to the next `sandboxir::Instruction`.

Reply:
oh I see. can you update the comments around `Val` (separate PR) to say that?